synapse/storage/databases/main/events_bg_updates.py

```diff
@@ -2429,6 +2435,113 @@ def _fill_table_txn(txn: LoggingTransaction) -> None:
 
         return len(memberships_to_update_rows)
 
+    async def _sliding_sync_membership_snapshots_fix_forgotten_column_bg_update(
+        self, progress: JsonDict, batch_size: int
+    ) -> int:
+        """
+        Background update to update the `sliding_sync_membership_snapshots` ->
+        `forgotten` column to be in sync with the `room_memberships` table.
+
+        Because of previously flawed code (now fixed); any room that someone has
+        forgotten and subsequently re-joined or had any new membership on, we need to go
+        and update the column to match the `room_memberships` table as it has fallen out
+        of sync.
+        """
+        last_event_stream_ordering = progress.get(
+            "last_event_stream_ordering", -(1 << 31)
+        )
+
+        # To simplify things, we can just recheck for any row in
+        # `sliding_sync_membership_snapshots` with `forgotten=1`
+        def _find_memberships_to_update_txn(
+            txn: LoggingTransaction,
+        ) -> List[Tuple[str, str, str, int]]:
+            txn.execute(
+                """
+                SELECT
+                    room_id,
+                    user_id,
+                    membership_event_id,
+                    event_stream_ordering
+                FROM sliding_sync_membership_snapshots
```

Comment:
May as well join against `room_memberships` so that we can see if the forgotten flags match? Rather than updating every row that has `forgotten` set.

```suggestion
                SELECT
                    s.room_id,
                    s.user_id,
                    s.membership_event_id,
                    s.event_stream_ordering,
                    m.forgotten
                FROM sliding_sync_membership_snapshots AS s
                INNER JOIN room_memberships AS m ON (m.membership_event_id = m.event_id)
```

Reply:
Good call :white_check_mark: 

Relying too much on the pattern for the other background update on `sliding_sync_membership_snapshots`